### PR TITLE
fix: fail closed on unclassified MCP tools

### LINKED ISSUE #181: Require explicit safety classification for auto-discovered MCP tools
## Summary

Change MCP auto-discovery so tools without safety annotations are not silently
classified as `SafetyClass.READ`. Unannotated tools should require an explicit
operator-supplied classification (or land in a restricted/quarantined class) before
they are invokable.

## Why this matters

`READ` is the most permissive class in the default policy: no role requirement, no
justification, highest rate limit. An MCP server exposing an unannotated destructive
tool (delete, transfer, send) is auto-registered with the weakest controls — the
opposite of least privilege, and silent. For a security kernel, insecure-by-default
classification of unknown tools is the highest-impact configuration risk.

## Current evidence

- `drivers/mcp.py:54-64`: `_infer_safety_class` returns `SafetyClass.READ` both for read-annotated tools and as the fallback for tools with no annotations; the docstring describes this as conservative, but `policy.py` treats `READ` as the least-restricted class (no role gate, highest default rate limit).
- `drivers/mcp.py:149` applies the inferred class during discovery with no override hook.

## External context

MCP tool annotations (`readOnlyHint`, `destructiveHint`) are advisory per the MCP
spec; gateway guidance recommends explicit allow-listing of tools rather than
trusting server-provided hints.

## Proposed implementation

1. Add a discovery policy parameter, e.g. `unannotated_safety: SafetyClass | "reject"`,
   defaulting to the most restrictive class (or rejection with a clear error listing
   the affected tools).
2. Provide a per-tool override map (`{tool_name: SafetyClass}`) so operators can
   classify deliberately.
3. Emit a structured warning listing every tool that fell through to the default.
4. Document that server-provided annotations are hints, not trusted statements.

## AI-agent execution notes

- Inspect first: `drivers/mcp.py` (`_infer_safety_class`, discovery loop), `drivers/mcp_support.py`, `enums.py` (`SafetyClass`), `policy.py` (per-class rules), `tests/test_mcp_driver.py`.
- Preserve behavior for tools with explicit annotations.
- Edge cases: server with zero annotated tools; annotation conflicts (read+destructive).
- Do not weaken the policy side; the change is in classification, not enforcement.

## Acceptance criteria

- An unannotated MCP tool is not registered as `READ` by default; it is restricted or rejected per configuration.
- Operators can override classification per tool, and overrides are tested.
- A warning names every tool that received the default classification.
- Existing annotated-tool behavior is unchanged.

## Test plan

Extend `tests/test_mcp_driver.py` with unannotated, mixed, and override scenarios;
assert resulting classes and warnings. Run `make ci`.

## Documentation plan

Update `docs/integrations.md` MCP section and `docs/security.md`; CHANGELOG `Changed`
with explicit upgrade note.

## Migration and compatibility notes

Breaking for deployments relying on auto-`READ`: provide
`unannotated_safety=SafetyClass.READ` as an explicit opt-back, and document the
one-line migration in the CHANGELOG and release notes.

## Risks and tradeoffs

Stricter defaults add onboarding friction for quick demos; mitigate with a clearly
documented opt-back and a helpful rejection message. Safety-first default fits the
project's stated purpose.

## Suggested labels

security, breaking-change, integrations

### PR DESCRIPTION
**Pull request by dgenio**
Closes #181.

## What changed

- `MCPDriver.discover()` no longer silently classifies a tool with missing safety metadata as `SafetyClass.READ`.
- Default discovery now rejects any tool that lacks both an explicit `safety_class_map` entry and a useful MCP safety hint.
- Operators can deliberately choose `unannotated_safety=SafetyClass.WRITE|DESTRUCTIVE|READ`; every tool receiving that fallback is named in a warning.
- Explicit operator classifications remain highest precedence; destructive MCP hints beat read-only hints.
- Classification logic lives in a small `mcp_classification.py` helper rather than inflating the driver module.
- Existing MCP tests now classify intentionally unannotated fixtures explicitly.
- Added focused fail-closed regression tests and a migration/security guide.

## Why

The previous fallback was opposite to least privilege: absence of MCP metadata became the least-restricted safety class. MCP annotations are advisory and missing metadata is not evidence that an operation is read-only.

## Compatibility

Deliberate breaking security default. Previous behavior can be requested explicitly:

```python
await driver.discover(unannotated_safety=SafetyClass.READ)
```

The recommended migration is an operator-reviewed `safety_class_map`.

## Security boundary

This fixes missing-metadata classification. It does **not** make server-provided annotations trustworthy; a dishonest server can still mislabel a tool. High-assurance deployments should use reviewed operator classification.

## Validation

CI adds tests for:
- default rejection + affected tool names;
- explicit per-tool classification;
- explicit fallback + warning containing all affected tools;
- read/destructive hint inference;
- invalid fallback configuration;
- existing discover/register/invoke and real FastMCP fixtures updated to classify intentionally.

This PR should merge independently of the broader MCP v2 migration in #263; #173 remains the real-server interoperability gate.

### CODE REVIEW

File: tests/test_mcp_discovery_safety.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        capabilities = await driver.discover(unannotated_safety=SafetyClass.WRITE)
+
+    assert {cap.safety_class for cap in capabilities} == {SafetyClass.WRITE}
+    assert "mcp_unannotated_safety_fallback" in caplog.text
+    assert "send_email" in caplog.text
+    assert "write_file" in caplog.text
+
+
+@pytest.mark.asyncio
+async def test_explicit_mcp_hints_still_infer_safety_class() -> None:
```
Comment (dgenio) [inline tests/test_mcp_discovery_safety.py:115]: The docs state that when both hints are present, destructive must win, but there is no regression test for the conflict case where readOnlyHint=true and destructiveHint=true on the same tool. The current tests still pass if precedence is accidentally inverted in a future refactor. Please add one focused test asserting the conflicting-hints tool is classified as DESTRUCTIVE.

File: docs/mcp-safety-classification.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# MCP tool safety classification
```
Comment (dgenio) [inline docs/mcp-safety-classification.md:1]: This new guide is useful, but it appears unreferenced from canonical docs surfaces. Please add links from docs/integrations.md and docs/security.md so migration/security guidance remains discoverable and consistent with the repo canonical-doc policy.

File: src/weaver_kernel/drivers/mcp_classification.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        names = ", ".join(sorted(rejected))
+        raise DriverError(
+            f"MCPDriver '{driver_id}' refused to auto-register unclassified tools: {names}. "
+            "Classify them with safety_class_map or explicitly set unannotated_safety to a "
+            "SafetyClass. MCP annotations are advisory and missing metadata is not treated "
+            "as READ authority."
+        )
+
+    if fallback:
+        assert isinstance(unannotated_safety, SafetyClass)  # narrowed above
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/drivers/mcp_classification.py]: Replaced `assert` with `typing.cast(SafetyClass, unannotated_safety)` for robust type narrowing under `-O`.

File: tests/test_mcp_discovery_safety.py (hunk 1 above)
Comment (dgenio) [reply tests/test_mcp_discovery_safety.py:115]: Added `test_conflicting_hints_prefers_destructive` asserting that a tool with both `readOnlyHint=True` and `destructiveHint=True` is classified as DESTRUCTIVE. Good catch — this was documented but untested.

File: src/weaver_kernel/drivers/mcp_classification.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        classified.append((spec, safety_class))
+
+    if rejected:
+        names = ", ".join(sorted(rejected))
+        raise DriverError(
+            f"MCPDriver '{driver_id}' refused to auto-register unclassified tools: {names}. "
+            "Classify them with safety_class_map or explicitly set unannotated_safety to a "
+            "SafetyClass. MCP annotations are advisory and missing metadata is not treated "
+            "as READ authority."
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/drivers/mcp_classification.py:81]: Fixed: message now reads "refused to discover/classify unclassified tools" to avoid conflating discovery with registration.

File: docs/mcp-safety-classification.md (hunk 2 above)
Comment (dgenio) [reply docs/mcp-safety-classification.md:1]: Added forward links from both `docs/integrations.md` (MCP Notes section) and `docs/security.md` (new MCP tool safety classification subsection).

File: src/weaver_kernel/drivers/mcp_classification.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    if fallback:
+        assert isinstance(unannotated_safety, SafetyClass)  # narrowed above
+        logger.warning(
+            "mcp_unannotated_safety_fallback driver_id=%s safety_class=%s tools=%s",
+            driver_id,
+            unannotated_safety.value,
+            ",".join(sorted(fallback)),
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/drivers/mcp_classification.py:90]: Changed to `%r` structured list (`sorted(fallback)`) replacing comma-joined string. Addresses comma-in-name ambiguity and improves log parsing.
